src/quantuminspire/qiskit/circuit_parser.py

```diff
@@ -130,101 +119,94 @@ def _z(self, circuit):
+            circuit (dict): The Qiskit circuit with U1 element.
 
         Returns:
-            Str: CQASM code string.
+            str: cQASM code string.
         """
-        angle_q0 = circuit['params'][0]
-        qubit_indices = tuple(circuit['qubits'])
-        operator = CircuitToString._u1_operator_switch.get(angle_q0)
-        parameters = circuit['params']
-        if operator is None:
-            raise ValueError('Gate U1 with parameters not implemented (parameters=%s)!' % parameters)
-        return '%s q[%d]\n' % (operator, *qubit_indices)
+        tempcircuit = circuit
+        tempcircuit['params'].insert(0, 0)
+        tempcircuit['params'].insert(0, 0)
+        tempcircuit['texparams'].insert(0, '0')
+        tempcircuit['texparams'].insert(0, '0')
+        return self._u3(tempcircuit)
 
     def _u2(self, circuit):
-        """ Translates the U2 element. Not usable!
+        """ Translates the U2(phi, lambda) element to U3(pi/2, phi, lambda).
 
         Args:
-            circuit (dict): The qiskit circuit with U2 element.
+            circuit (dict): The Qiskit circuit with U2 element.
 
-        Raises:
-            ValueError: When the provided rotation angles are invalid!
         """
-        parameters = circuit['params']
-        raise ValueError('Gate U2 not implemented (parameters=%s)!' % parameters)
+        tempcircuit = circuit
+        tempcircuit['params'].insert(0, np.pi/2)
+        tempcircuit['texparams'].insert(0, '\\frac{\\pi}{2}')
+        return self._u3(tempcircuit)
 
     def _u3(self, circuit):
-        """ Translates the U3 element.
+        """ Translates the U3(theta, phi, lambda) element to 3 rotation gates.
+            Any single qubit operation (a 2x2 unitary matrix) can be written as the product of rotations.
+            As an example, a unitary single-qubit gate can be expressed as a combination of
+            Rz and Ry rotations (Nielsen and Chuang, 10th edition, section 4.2).
+            U(theta, phi, lambda) = Rz(phi)Ry(theta)Rz(lambda).
+            Note: The expression above is the matrix multiplication, when implementing this in a gate circuit,
+            the gates need to be executed in reversed order.
+            Any rotation of 0 radials is left out of the resulting circuit.
 
         Args:
-            circuit (dict): The qiskit circuit with U3 element.
+            circuit (dict): The Qiskit circuit with U3 element.
 
         Returns:
-            Str: CQASM code string.
+            str: cQASM code string.
         """
         gates = ['Rz', 'Ry', 'Rz']
-        angles = circuit['params'][:3]
+        angles = list(circuit['params'][i] for i in [2, 0, 1])
         index_q0 = [circuit['qubits'][0]] * 3
-        return ''.join('%s q[%d], %f\n' % pair for pair in zip(gates, index_q0, angles))
+        return ''.join('%s q[%d], %f\n' % triplet for triplet in zip(gates, index_q0, angles) if triplet[2] != 0)
```

Comment:
If theta is zero, then the output is two Rz gates, not one, is that something that matters?

Reply:
Nope, it's what is intended. Maybe an extra optimization is possible here. But when theta is zero, the quantum programmer shouldn't use the u3 gate but an Rz gate instead.

Reply:
✔️